pkg/processor/runtime/python/py/test_server_wrapper.py

```diff
@@ -0,0 +1,218 @@
+        self._host = "0.0.0.0"
+        self._port = 1337
+
+        self._wrapper = wrapper.Wrapper(
+            logger=self._logger,
+            loop=self._loop,
+            handler=self._default_test_handler,
+            serving_address=f"{self._host}:{self._port}",
+            control_socket_path=None,
+            platform_kind=self._platform_kind,
+            decode_event_strings=self._decode_event_strings)
+        self._loop.run_until_complete(self._wrapper.initialize())
+        self._wrapper_run_task = self._loop.create_task(self._wrapper.start())
+
+    def tearDown(self):
+        sys.path.remove(self._temp_path)
+        asyncio.run(self._wrapper._stop_processing())
+        self._wrapper_run_task.cancel()
+        self._loop.close()
+
+    def test_async_handler_single_connection(self):
+        self._test_async_handler(single_connection=True)
+
+    def test_async_handler_multiple_connections(self):
+        self._test_async_handler(single_connection=False)
+
+    def _test_async_handler(self, single_connection):
+        """Test function decorated with async and running an event loop"""
+
+        recorded_events = []
+
+        async def event_recorder(context, event):
+            # random execution time
+            await asyncio.sleep(random.uniform(0.1, 1))
+            recorded_events.append(event)
+            context.logger.debug_with('appending event', event=repr(event.id))
+            return 'ok'
+
+        num_of_events = 10
+        events = (
+            nuclio_sdk.Event(_id=i, body='e{}'.format(i))
+            for i in range(num_of_events)
+        )
+        self._wrapper._entrypoint = event_recorder
+        self._loop.run_until_complete(self._send_events(events, single_connection=single_connection))
+
+        self.assertEqual(num_of_events, len(recorded_events), 'wrong number of events')
+        if single_connection:
+            # we expect the event to be ordered since though the function is "asynchronous", it is blocked
+            # by the processor until it gets response.
+            for recorded_event_index, recorded_event in enumerate(
+                    sorted(recorded_events, key=operator.attrgetter('id'))):
+                self.assertEqual(recorded_event_index, recorded_event.id)
+                self.assertEqual('e{}'.format(recorded_event_index), self._ensure_str(recorded_event.body))
+        else:
+            expected_events = [
+                {'id': i, 'body': f'e{i}'}
+                for i in range(num_of_events)
+            ]
+            actual_events = [
+                {'id': recorded_event.id, 'body': self._ensure_str(recorded_event.body)}
+                for recorded_event in recorded_events
+            ]
+            self.assertEqual(
+                collections.Counter(map(frozenset, expected_events)),
+                collections.Counter(map(frozenset, actual_events)),
+                "Recorded events do not match the expected events"
+            )
```

Comment:
This only verifies that all events arrived, right?
How is this testing asynchronicity?

Reply:
@TomerShor just see the logs to check async. The test itself doesn't test async, it tests that all events arrived. Technically, we could add the check that they are not ordered, however there is still a small chance that they will be (because sleep time is random)

Reply:
Yep, we can't guarantee it really.